File: 98-validate-binary-search-tree/98-validate-binary-search-tree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def isValidBST(root: TreeNode): Boolean = {
        validate(root)._3
    }
    
    def validate(root: TreeNode): (Long, Long, Boolean) = {
        if (root == null) {
            (Long.MaxValue, Long.MinValue, true)
        } else {
            val (lmin, lmax, lvld) = validate(root.left)
            val (rmin, rmax, rvld) = validate(root.right)
            (lmin min root.value, rmax max root.value,  root.value > lmax && root.value < rmin && lvld && rvld)
        }
    }
}

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return validateHelper(root, null, null);
    }
    
    boolean validateHelper(TreeNode root, Integer min, Integer max) {
        if (root == null) return true;
        
        if ((min != null && root.val <= min) || (max != null && root.val >= max)) return false;
        
        return validateHelper(root.left, min, (max == null)? root.val : Math.min(root.val, max)) &&
            validateHelper(root.right, (min == null)? root.val : Math.max(root.val, min), max);
    }
}